### PR TITLE
Fix `ZeroDivisionError` on 0-row Parquets with boolean cols, upgrade to Python 3.11

### PR DESCRIPTION
**Pull request by ryan-williams**
## Before
`ZeroDivisionError` on 0-row `test0.parquet` that contains a `bool` col:

<details><summary><code>parquet-tools inspect ~/c/parquet-tools/tests/test0.parquet</code></summary>

```
############ file meta data ############
created_by: parquet-cpp-arrow version 14.0.2
num_columns: 4
num_rows: 0
num_row_groups: 1
format_version: 2.6
serialized_size: 2398

Traceback (most recent call last):
  File "/Users/ryan/.pyenv/versions/parquet-diff-test-3.11.6/bin/parquet-tools", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Users/ryan/.pyenv/versions/3.11.6/envs/parquet-diff-test-3.11.6/lib/python3.11/site-packages/parquet_tools/cli.py", line 26, in main
    args.handler(args)
  File "/Users/ryan/.pyenv/versions/3.11.6/envs/parquet-diff-test-3.11.6/lib/python3.11/site-packages/parquet_tools/commands/inspect.py", line 54, in _cli
    _execute_simple(
  File "/Users/ryan/.pyenv/versions/3.11.6/envs/parquet-diff-test-3.11.6/lib/python3.11/site-packages/parquet_tools/commands/inspect.py", line 66, in _execute_simple
    print(_simple_schema_expression(file_meta, file_schema))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/.pyenv/versions/3.11.6/envs/parquet-diff-test-3.11.6/lib/python3.11/site-packages/parquet_tools/commands/inspect.py", line 92, in _simple_schema_expression
    col_compression_space_saving_ratio = 1 - (col_meta.total_compressed_size / col_meta.total_uncompressed_size)
                                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ZeroDivisionError: division by zero
```
</details> 

## After
Succeeds + prints `N/A` (see also `test0_inspect.txt`):

<details><summary><code>parquet-tools inspect ~/c/parquet-tools/tests/test0.parquet</code></summary>

```

############ file meta data ############
created_by: parquet-cpp-arrow version 14.0.2
num_columns: 4
num_rows: 0
num_row_groups: 1
format_version: 2.6
serialized_size: 2398


############ Columns ############
a
b
c
d

############ Column(a) ############
name: a
path: a
max_definition_level: 1
max_repetition_level: 0
physical_type: DOUBLE
logical_type: None
converted_type (legacy): NONE
compression: SNAPPY (space_saved: -7%)

############ Column(b) ############
name: b
path: b
max_definition_level: 1
max_repetition_level: 0
physical_type: INT32
logical_type: Null
converted_type (legacy): NONE
compression: SNAPPY (space_saved: -7%)

############ Column(c) ############
name: c
path: c
max_definition_level: 1
max_repetition_level: 0
physical_type: INT64
logical_type: None
converted_type (legacy): NONE
compression: SNAPPY (space_saved: -7%)

############ Column(d) ############
name: d
path: d
max_definition_level: 1
max_repetition_level: 0
physical_type: BOOLEAN
logical_type: None
converted_type (legacy): NONE
compression: SNAPPY (space_saved: N/A)

```
</details>

## Upgrade to Python 3.11
3.8 is [nearing EoL](https://devguide.python.org/versions/) (fall 2024), so this PR also:
- upgrades CI to use 3.11
- bumps (and pins) various dep versions
- adds a `write_test_dataframes` helper that generates the `tests/test*.parquet` files
- uses it to regenerate `test{1,2}.parquet`

### CODE REVIEW

**Review by ktrueda — COMMENTED** [review]
Thank you for making PR. Could you confirm the version expressions in pyproject.toml ?

**Comment by ktrueda** [comment]
I found a weird behavior thanks to this CI.
https://github.com/ktrueda/parquet-tools/pull/50/checks
image

If users specify multiple parquet files which has different column structure, the outcome is concat with column direction (not row direction). As this issue has nothing to do with this PR, I'll make a issue after merging this PR.

File: pyproject.toml [hunk 1]
```diff
@@ -11,25 +11,25 @@ keywords = ["parquet-tools", "parquet"]
...
-thrift = "^0.13.0"
-colorama = "^0.4.3"
+python = "^3.11"
+halo = "0.0.31"
+pyarrow = "14.0.2"
+pandas = "2.1.4"
+tabulate = "0.9.0"
+boto3 = "1.34.11"
+thrift = "0.16.0"
+colorama = "0.4.6"
```
[the remark this answers is not in the document]
Comment (ryan-williams) [reply pyproject.toml]: Good point. I've added a commit relaxing these versions. I've made them all `*` (to hopefully not interfere with any existing versions in a user's environment), with two exceptions:

- The full pyarrow version appears in the test Parquet files / expected test data, so I've pinned that exactly in the `dev-dependencies`.
- `python = ">=3.8"`

Let me know if you think just preprending `^` to the previous versions is better, though.

[the remark this answers is not in the document]
Comment (ktrueda) [reply pyproject.toml]: It's difficult to decide 🤔 

As we run some testw with specific versions, we can't guarantee other versions. So I think we should prepend `^` to versions for dev-dependencies.

[the remark this answers is not in the document]
Comment (ryan-williams) [reply pyproject.toml]: Sounds good, I just pushed a commit with all versions beginning with `^`, except `python>=3.8` and `pyarrow==14.0.2` (in dev-dependencies only, `*` in regular dependencies).

[the remark this answers is not in the document]
Comment (ktrueda) [reply pyproject.toml]: Thank you for modifying. It looks good.

**Review by ktrueda — APPROVED** [review]
LGTM

**Pull request merged**